### DISCUSSION TITLE
Subtitle colors?

### DISCUSSION THREAD

**Discussion by Lexx2k** [sfall-team/sfall#372] [General]
This is a bit of an odd question, but it is driving me insane for the past 4 days now ...

... How do the subtitle colors in the endgame slideshow images work? It doesn't make sense to me. I am trying to create a custom .pal file, but the colors never really match. It seems that index 255 is white = font. Though, that isn't the case in every ending slide image?

Here are two examples:
![subs_1](https://user-images.githubusercontent.com/4844938/116515495-a9c53300-a8cc-11eb-9a06-38ad2819a703.png)

![subs_2](https://user-images.githubusercontent.com/4844938/116515507-acc02380-a8cc-11eb-9ea8-d3fbb6af5b38.png)

And here is my file:
![subs_3](https://user-images.githubusercontent.com/4844938/116515521-b184d780-a8cc-11eb-8e67-856079431669.png)

^as you can see, the background color is not black, and the font blending is not working correctly.

Another example:
![subs_4](https://user-images.githubusercontent.com/4844938/116515862-235d2100-a8cd-11eb-95ea-efaf186befa9.png)

Basically, the only thing I found out is that index 255 is the flat font color, but as can be seen on my second image above (followers ending), color index 255 is *not* white there, which means this is not necessarily true.

Anyways... with 255 working as font color, what's missing is the actual background color, and the blending colors ... I have no idea on what index they are. They don't seem to be in order? Or does the game not check for the color index and instead picks the color that is the closest to what it needs?

Note:
As a random test, I made a screenshot of the working subtitles ingame, then copy&pasted the text part into my new image, and then converted it to indexed colors so that my new .pal file has the necessary colors. This experiment didn't work out ... which tells me that the colors must be on a specific index.

**Comment by FakelsHub** [comment]
From what I understand, the Subtitles should always be white. And the palette should always contain white and black color.
it's all very complicated with the font color.

**Reply by Lexx2k** [reply]
Here is what I found out so far:
In all of my tests, index 255 is the white font color and index 179 is the black background. If I change index 179 to a different color, then that will be the subtitle background. If I add these colors on any other index in the palette, it will not work / the game keeps picking the specific color index.

However, the text blending will still not work correctly - I am assuming there are a lot more colors needed for that. It's just impossible tedious to figure that out with try&error.

But ... this makes no sense, because in the official ending slides, black is *not* always on index 179. 

Could it be that there is specific subtitle information saved in the fallout .pal files?

**Reply by FakelsHub** [reply]
There are no specific indexes in the engine code, the game itself somehow chooses the colors of the indexes. That's why I pointed out that everything is difficult there with these. We need to carefully understand the algorithms of the work.

**Reply by Lexx2k** [reply]
In any case, I have the same effects in every image I tried, even if they have a different color scheme. It's always the same.

Maybe it has something to do with how photoshop (can't test with anymore else) is sorting the colors in the palette? Because if I apply a vanilla .pal file to an image, the subtitles work.

**Reply by FakelsHub** [reply]
see which index has the white color in the default game palette (color.pal)
. perhaps this may be related.

**Reply by Lexx2k** [reply]
[color.pal](https://fodev.net/files/fo2/fo_colors.html#index220) has no pure white. The closest is 252, 252, 252 on index 220.

**Reply by FakelsHub** [reply]
somewhere there was already a discussion of the colors of subtitles, need to find this among the more commits)

**Reply by fedprod** [reply]
subtitle.pal https://i.imgur.com/6qz4CAI.png

**Reply by Lexx2k** [reply]
Not every slideshow image has all these colors, see this as example:
![subs_5](https://user-images.githubusercontent.com/4844938/116606977-bdef4b80-a931-11eb-9c36-65c9b9ed8e05.png)

/Edit: Alright, this one has some non-gray blending colors. Still, the question is how the game detects what color to use for subtitles, because in this image, the index for black (as example) is 1 and 254.
![subs_5](https://user-images.githubusercontent.com/4844938/116607702-9a78d080-a932-11eb-89b3-b5b3cf26ddd8.png)

**Reply by FakelsHub** [reply]
I once again looked at the implementation of the functions, from the defined it is **white color of the subtitles** and **black color of the background**.
Color indexes from the palette for text and blending color are selected by the algorithms of the game function (these functions are quite incomprehensible, they have a lot of arithmetic.)

**Reply by Lexx2k** [reply]
I just noticed something interesting with the endgame slides: I had issues with one slide, then I changed the appearance order of the slides (endgame.txt list) ... and suddenly the issue was gone (I'm trying to apply vanilla .pal files now, with mixed results).

Looks to me like the last loaded .pal in the slideshow also has an influence on what comes after it. Or this is just a bug due to the fucked up colors.

**Reply by Lexx2k** [reply]
Another random idea:
What if ending slides would get a similar treatment as the 32bit talking heads? Basically rewriting the function to show 32bit images, with the subtitle colors always being in a specific place. This way nobody needs to figure out how the hell the original subtitle algorithm works and we would not have to bother with the limited colors anymore ...

**Reply by FakelsHub** [reply]
What is your problem?

**Reply by Lexx2k** [reply]
The problem is that I can't get the subtitles to work without shit fucking up. :D

Even if I try to use one of the original .pal files, half the time it just fails.

**Reply by FakelsHub** [reply]
Show your problem. :-)

**Reply by Lexx2k** [reply]
Uhh.... my problem can be seen all over this thread? Like, already in the very first post?

But ok, here it is again: https://imgur.com/a/rXqY989

Here is the thing, though. The first and second image *already worked before*. I don't know why they stopped working. 
The 3rd image is vanilla and I didn't touch it at all -- but suddenly you can see these black artifacts on the armor. Similar happens with Junktown now.
So it seems that whatever is doing the subtitle shit not only affects new images, but also old images.

**Reply by FakelsHub** [reply]
check has the black color is 0.0.0 in the palette.

**Reply by Lexx2k** [reply]
First two images are using the vanilla hub .pal file (I gave up trying to make new .pal files for now) and for both, black is on index 179.
I wrote that already further above.

**Reply by FakelsHub** [reply]
About the black spots. Check out the w/o hrp and disable sfall dx9 mode

**Reply by Lexx2k** [reply]
I tried with mode 0 and it was still the same. After removing the HRP the issue was gone.

PS: As I wrote before, subtitle / image colors can break depending on what ending slide was playing *before*. For example everything works if I have image A for the Hub, but all following images will break if I am showing image B for the Hub instead.

**Reply by FakelsHub** [reply]
> As I wrote before, subtitle / image colors can break depending on what ending slide was playing before.

I think I found a bug with the palette there, could you give me a savegame for testing subtitles.

**Reply by Lexx2k** [reply]
You will have to grab the latest files from the Fo1in2 repo, for the new graphics, etc.

Here is the savegame: 
[SLOT01.zip](https://github.com/phobos2077/sfall/files/6445816/SLOT01.zip)

You also need this script file:
[V13CAVE.zip](https://github.com/phobos2077/sfall/files/6445818/V13CAVE.zip)

Just leave the V13 cave, then skip the cutscene and go back to the V13 cave. The ending sequence will start. To skip the cutscene with the overseer at the end, be sure to use alt+f4 before getting there.

It seems the ending slides start bugging after the Brotherhood image. Because everything before works fine.

**Reply by Lexx2k** [reply]
I just noticed that even the outro cutscene (walking into the desert) has broken colors in that ending slide constellation.

**Reply by FakelsHub** [reply]
I won't be doing this any time soon. Now we need to prepare for the release of sfall.

**Reply by FakelsHub** [reply]
> Just leave the V13 cave, then skip the cutscene and go back to the V13 cave. The ending sequence will start.

nothing starts!
1. leave -> enter
2. the dialogue with the overseer begins after the conversation, the video of the chosen one's departure to the desert - > starts credits, then the menu pops up you want to continue playing yes/no and no end slides.

**Reply by Lexx2k** [reply]
Are you using the latest version? Because I disabled the extended endings before last release, which now will likely fuck with the savegame.

**Reply by FakelsHub** [reply]
> Are you using the latest version?

no, i have v1.3. why do I need to use the latest version?)
you definitely have everything in order there? because the final slides are simply skipped when calling the function (it looks like a bug)
```
      gfade_out(1);
      endgame_slideshow; 
```

**Reply by Lexx2k** [reply]
Well I don't remember, it's been months. Back then when I was showcasing it, it was all working like I described it. Sadly I don't have time right now to dive into it again and investigate what's going on. Maybe in a month or two.

**Reply by Lexx2k** [reply]
> no, i have v1.3.

Actually this might be the problem. The new ending slides were added much later... So if you use the savegame and script with that old version, of course it will skip the ending slides.. because it can't find them (endgame.txt, interface files, etc).

**Reply by FakelsHub** [reply]
what version do I need for this to work?

**Reply by Lexx2k** [reply]
Latest repo version + the v13cave.int file from further above should work. The edited v13cave.int file will always trigger the expanded ending slides, which is where the subtitle bug is most noticeable.

**Reply by FakelsHub** [reply]
There is no bugs, you just don't know how to use the palette. ;p

**Reply by FakelsHub** [reply]
> Latest repo version

I understand it, it is enough to replace FO1_BASE with a new one and not bother with any installation?

**Reply by FakelsHub** [reply]
Ok. I made it so in the script that it works when the new game start.
```
//      if (((global_var(1101) >= 2) andAlso (((global_var(1180) > 0) andAlso (global_var(1316) > 0)) andAlso (endgame_running == 0))) andAlso not(global_var(1618) > 0)) then begin
//         game_ui_disable;
call Endgame_Slides_Expanded_Debug();
//      end
//      else begin
//         override_map_start(136, 129, 0, 1);
//      end
```
I saw some problematic subtitles. Expect a lesson on preparing a palette. :)

**Reply by FakelsHub** [reply]
colors in the palette 
1 the index must be black 0.0.0
255 the index must be white.
there is still a problem of mixing colors, probably the engine's algorite itself does not work well, or it lacks the required colors in the palette.
I suspect that the colors should be located from dark (0-index) to light (255-index) then mixing will probably work.

I will continue the tests later.

**Comment by egornovivan** [comment]
Turn off windows aero and check again

**Reply by Lexx2k** [reply]
I have aero and all that jazz disabled by default.
/Edit: Actually Win10 doesn't even have that.

**Comment by FakelsHub** [comment]
Use it: https://github.com/FakelsHub/FixPAL
This will fix your wrong palette.
Before creating a palette for your image, the palette should contain black and white color no matter where.
If the image does not have such colors, you must reserve two free spaces in order to manually assign these colors to them later.
Good Luck :)
